File: examples/python/simple_workflow.py

import os
import shutil
from kamaros import JCFManager, FileAdapter

def main():
    # Setup paths
    project_path = "./demo-project-store"
    if os.path.exists(project_path):
        shutil.rmtree(project_path)
    os.makedirs(project_path)

    print(f"--- Kamaros Python Demo ---")
    print(f"Storage path: {project_path}")

    # 1. Initialize Manager
    adapter = FileAdapter(project_path)
    manager = JCFManager(adapter)
    
    # 2. Create Project
    print("\n[1] Creating project 'DemoApp'...")
    manager.create_project("DemoApp", author="Python Example")
    
    # 3. Add initial file
    print("[2] Adding 'README.md'...")
    manager.add_file("README.md", b"# Demo Project\nInitial content.")
    
    # 4. Save Checkpoint v1
    v1_id = manager.save_checkpoint("Initial commit")
    print(f" -> Checkpoint saved: {v1_id}")

    # 5. Modify file
    print("\n[3] Modifying 'README.md'...")
    manager.add_file("README.md", b"# Demo Project\nUpdated content with new features.")
    
    # 6. Save Checkpoint v2
    v2_id = manager.save_checkpoint("Update README")
    print(f" -> Checkpoint saved: {v2_id}")

    # Verify current state
    current_content = manager.get_file("README.md")
    print(f"Current content: {current_content.decode('utf-8').strip()}")
    assert b"Updated content" in current_content

    # 7. Restore v1
    print(f"\n[4] Restoring version {v1_id}...")
    restored_id = manager.restore_version(v1_id)
    print(f" -> Restored to: {restored_id}")

    # 8. Verify restoration
    restored_content = manager.get_file("README.md")
    print(f"Restored content: {restored_content.decode('utf-8').strip()}")
    
    if b"Initial content" in restored_content:
        print("\nSUCCESS: Content restored correctly!")
    else:
        print("\nFAILURE: Content mismatch!")

if __name__ == "__main__":
    main()

File: examples/ts/simple-workflow.ts

import { JCFManager, NodeAdapter, initWasm } from '../../js/src/index.ts';
import * as fs from 'fs';
import * as path from 'path';

async function main() {
    // Navigate to project root to find wasm pkg if needed, or rely on correct relative paths
    // In this dev setup, we are inside examples/ts.

    const projectPath = path.resolve('./demo-project-store-ts');

    // Cleanup
    if (fs.existsSync(projectPath)) {
        fs.rmSync(projectPath, { recursive: true, force: true });
    }
    fs.mkdirSync(projectPath);

    console.log("--- Kamaros TypeScript Demo ---");
    console.log(`Storage path: ${projectPath}`);

    // 0. Init WASM (required for TS/Browser)
    console.log("\n[0] Initializing WASM...");
    await initWasm();

    // 1. Initialize Manager
    const adapter = new NodeAdapter(projectPath);
    const manager = await JCFManager.create(adapter);

    // 2. Create Project
    console.log("\n[1] Creating project 'DemoAppTS'...");
    await manager.createProject("DemoAppTS");

    // 3. Add initial file
    console.log("[2] Adding 'index.ts'...");
    const initialContent = new TextEncoder().encode("console.log('Hello v1');");
    await manager.addFile("index.ts", initialContent);

    // 4. Save Checkpoint v1
    const v1Id = await manager.saveCheckpoint("Initial commit TS");
    console.log(` -> Checkpoint saved: ${v1Id}`);

    // 5. Modify file
    console.log("\n[3] Modifying 'index.ts'...");
    const updatedContent = new TextEncoder().encode("console.log('Hello v2 - updated');");
    await manager.addFile("index.ts", updatedContent);

    // 6. Save Checkpoint v2
    const v2Id = await manager.saveCheckpoint("Update index.ts");
    console.log(` -> Checkpoint saved: ${v2Id}`);

    // Verify current
    const currentFile = await manager.getFile("index.ts");
    const currentStr = new TextDecoder().decode(currentFile);
    console.log(`Current content: ${currentStr}`);
    if (!currentStr.includes("v2")) throw new Error("Verification failed: expected v2 content");

    // 7. Restore v1
    console.log(`\n[4] Restoring version ${v1Id}...`);
    await manager.restoreVersion(v1Id);
    console.log(` -> Restored to: ${v1Id}`);

    // 8. Verify restoration
    const restoredFile = await manager.getFile("index.ts");
    const restoredStr = new TextDecoder().decode(restoredFile);
    console.log(`Restored content: ${restoredStr}`);

    if (restoredStr.includes("v1")) {
        console.log("\nSUCCESS: Content restored correctly!");
    } else {
        console.error("\nFAILURE: Content mismatch!");
        process.exit(1);
    }
}

main().catch(err => {
    console.error("Error:", err);
    process.exit(1);
});
